src/Microsoft.AspNet.Mvc.Core/Internal/Routing/LinkGenerationDecisionTree.cs

```diff
@@ -131,19 +133,36 @@ private class LinkGenerationMatchComparer : IComparer<LinkGenerationMatch>
 
             public int Compare(LinkGenerationMatch x, LinkGenerationMatch y)
             {
-                // For these comparisons lower is better.
-
+                //For this comparison lower is better
                 if (x.Entry.Order != y.Entry.Order)
                 {
                     return x.Entry.Order.CompareTo(y.Entry.Order);
                 }
 
+                if(x.ParameterCount != y.ParameterCount)
+                {
+                    if(x.Entry?.Template.Segments?.Count != y.Entry?.Template.Segments?.Count)
```

Comment:
What's happening here? Why does `x.ParameterCount.CompareTo(y.ParameterCount)` not suffice?